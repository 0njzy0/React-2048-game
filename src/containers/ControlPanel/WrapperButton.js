import React from 'react';
import PropTypes from 'prop-types';
import styles from './controlPanel.scss';
import Button from '../../components/Button';

export default function WrapperButton({ children, onClick, ...props }) {
  return (
    <div className={styles.btn} >
      <Button onClick={onClick} {...props}>
        {
          children
        }
      </Button>
    </div>
  );
}

WrapperButton.propTypes = {
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
};
