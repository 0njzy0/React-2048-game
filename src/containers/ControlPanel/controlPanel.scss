@import '../../assets/styles/vars';

$border: 10px solid $btn-color;
$trans-border: 10px solid transparent;

.panel {
  align-items: center;
  display: flex;
  flex: 1;
  flex-direction: column;
  margin-bottom: 2px;

  .row-btn {
    align-items: center;
    display: flex;
    flex-grow: 1;
  }

  .btn {
    margin-right: 10px;
    margin-top: 10px;
  }

  img {
    margin-top: 4px;
    width: 27px;
  }

  .arrows {
    display: flex;
    flex-direction: row;
  }

  .left {
    transform: rotate(270deg);
  }

  .right {
    transform: rotate(90deg);
  }

  .down {
    transform: rotate(180deg);
  }
}
