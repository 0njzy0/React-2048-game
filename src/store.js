import { createStore } from 'redux';
import undoable from 'redux-undo';
import rootReducer from './reducers';

const initHistory = JSON.parse(localStorage.getItem('state') || 'null');
/* eslint-disable no-underscore-dangle */
const args = [
  undoable(rootReducer, {
    limit: 11, // set a limit for the history
    ignoreInitialState: true,
  }),
  window.__REDUX_DEVTOOLS_EXTENSION__ &&
  window.__REDUX_DEVTOOLS_EXTENSION__(),
];
if (initHistory) {
  args.splice(1, 0, initHistory);
}

const store = createStore(...args);

store.subscribe(() => {
  const state = store.getState();
  localStorage.setItem('state', JSON.stringify(state));
});

export default store;
