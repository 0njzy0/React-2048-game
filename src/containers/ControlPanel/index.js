import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import debounce from 'lodash.debounce';
import WrapperButton from './WrapperButton';
import Speaker from '../Speaker';
import styles from './controlPanel.scss';
import {
  moveUp,
  moveDown,
  moveLeft,
  moveRight,
  placeRandom,
  reset,
  revert,
} from '../../reducers/board';
import resetSvg from '../../assets/svg/reset.svg';
import undoSvg from '../../assets/svg/undo.svg';
import arrowSvg from '../../assets/svg/arrow.svg';
import moveAudio from '../../assets/audio/move.mp3';
import popupAudio from '../../assets/audio/popup.mp3';

class ControlPanel extends Component {
  static propTypes = {
    delay: PropTypes.number.isRequired,
    onMoveUp: PropTypes.func.isRequired,
    onMoveDown: PropTypes.func.isRequired,
    onMoveLeft: PropTypes.func.isRequired,
    onMoveRight: PropTypes.func.isRequired,
    onPlaceRandom: PropTypes.func.isRequired,
    onReset: PropTypes.func.isRequired,
    onRevert: PropTypes.func.isRequired,
    isMoved: PropTypes.bool.isRequired,
  }

  constructor(...args) {
    super(...args);

    this.audioMove = new Audio(moveAudio);
    this.audioPopup = new Audio(popupAudio);

    this.handleMoveUp = this.handleMoveUp.bind(this);
    this.handleMoveDown = this.handleMoveDown.bind(this);
    this.handleMoveLeft = this.handleMoveLeft.bind(this);
    this.handleMoveRight = this.handleMoveRight.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleSpeakerClick = this.handleSpeakerClick.bind(this);

    this.state = {
      speakerOn: true,
    };
  }

  componentWillMount() {
    const { delay } = this.props;
    document.body.addEventListener('keyup',
      debounce(this.handleKeyDown, delay, {
        leading: true,
      }));
  }

  handleKeyDown(e) {
    const keyUp = 38;
    const keyRight = 39;
    const keyDown = 40;
    const keyLeft = 37;
    const keyW = 87;
    const keyS = 83;
    const keyA = 65;
    const keyD = 68;
    const keyN = 78;

    switch (e.keyCode) {
      case keyW:
      case keyUp:
        this.handleMoveUp();
        break;
      case keyS:
      case keyDown:
        this.handleMoveDown();
        break;
      case keyA:
      case keyLeft:
        this.handleMoveLeft();
        break;
      case keyD:
      case keyRight:
        this.handleMoveRight();
        break;
      case keyN:
        this.props.onReset();
        break;
      default:
        break;
    }
  }

  handleSpeakerClick(speakerOn) {
    this.setState({
      speakerOn,
    });
  }

  generalMove(func) {
    func();
    const { isMoved } = this.props;
    const { speakerOn } = this.state;
    if (speakerOn && isMoved) {
      this.audioMove.play();
    }
    setTimeout(() => {
      this.props.onPlaceRandom();
      if (speakerOn && isMoved) {
        this.audioPopup.play();
      }
    }, 300);
  }

  handleMoveUp() {
    return this.generalMove(this.props.onMoveUp);
  }

  handleMoveDown() {
    this.generalMove(this.props.onMoveDown);
  }

  handleMoveLeft() {
    this.generalMove(this.props.onMoveLeft);
  }

  handleMoveRight() {
    this.generalMove(this.props.onMoveRight);
  }

  render() {
    const { delay } = this.props;
    return (
      <div className={styles.panel}>
        <div className={styles.rowBtn}>
          <div className={styles.btn}>
            <Speaker onClick={this.handleSpeakerClick} />
          </div>
          <WrapperButton onClick={debounce(this.props.onRevert, delay)} >
            <img src={undoSvg} alt="undo" />
          </WrapperButton>
          <WrapperButton onClick={debounce(this.props.onReset, delay)} >
            <img src={resetSvg} alt="reset" />
          </WrapperButton>
        </div>
        <WrapperButton type="primary" onClick={debounce(this.handleMoveUp, delay)} >
          <img src={arrowSvg} className={styles.up} alt="arrow up" />
        </WrapperButton>
        <div className={styles.arrows} >
          <WrapperButton type="primary" onClick={debounce(this.handleMoveLeft, delay)} >
            <img src={arrowSvg} className={styles.left} alt="arrow left" />
          </WrapperButton>
          <WrapperButton type="primary" onClick={debounce(this.handleMoveDown, delay)} >
            <img src={arrowSvg} className={styles.down} alt="arrow down" />
          </WrapperButton>
          <WrapperButton type="primary" onClick={debounce(this.handleMoveRight, delay)} >
            <img src={arrowSvg} className={styles.right} alt="arrow right" />
          </WrapperButton>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  isMoved: state.board.isMoved,
});
const mapDispatchToProps = dispatch => ({
  onPlaceRandom() {
    dispatch(placeRandom());
  },
  onMoveUp() {
    dispatch(moveUp());
  },
  onMoveDown() {
    dispatch(moveDown());
  },
  onMoveLeft() {
    dispatch(moveLeft());
  },
  onMoveRight() {
    dispatch(moveRight());
  },
  onReset() {
    dispatch(reset());
  },
  onRevert() {
    dispatch(revert());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ControlPanel);
