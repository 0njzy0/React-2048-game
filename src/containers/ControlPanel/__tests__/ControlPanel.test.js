import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import ControlPanel from '..';
import store from '../../../store';

describe('<ControlPanel />', () => {
  it('component render', () => {
    const panel = renderer.create(
      <Provider store={store}>
        <ControlPanel delay={1} />
      </Provider>,
    ).toJSON();
    expect(panel).toMatchSnapshot();
  });
});
