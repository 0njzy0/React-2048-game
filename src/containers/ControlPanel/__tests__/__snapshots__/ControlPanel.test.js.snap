// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ControlPanel /> component render 1`] = `
<div
  className="panel"
>
  <div
    className="rowBtn"
  >
    <div
      className="btn"
    >
      <button
        className="btnFlat btnMd btnDefault"
        onClick={[Function]}
      >
        <img
          alt="speaker controller"
          src="speaker-on.svg"
        />
      </button>
    </div>
    <div
      className="btn"
    >
      <button
        className="btnFlat btnMd btnDefault"
        onClick={[Function]}
      >
        <img
          alt="undo"
          src="undo.svg"
        />
      </button>
    </div>
    <div
      className="btn"
    >
      <button
        className="btnFlat btnMd btnDefault"
        onClick={[Function]}
      >
        <img
          alt="reset"
          src="reset.svg"
        />
      </button>
    </div>
  </div>
  <div
    className="btn"
  >
    <button
      className="btnFlat btnMd btnPrimary"
      onClick={[Function]}
    >
      <img
        alt="arrow up"
        className="up"
        src="arrow.svg"
      />
    </button>
  </div>
  <div
    className="arrows"
  >
    <div
      className="btn"
    >
      <button
        className="btnFlat btnMd btnPrimary"
        onClick={[Function]}
      >
        <img
          alt="arrow left"
          className="left"
          src="arrow.svg"
        />
      </button>
    </div>
    <div
      className="btn"
    >
      <button
        className="btnFlat btnMd btnPrimary"
        onClick={[Function]}
      >
        <img
          alt="arrow down"
          className="down"
          src="arrow.svg"
        />
      </button>
    </div>
    <div
      className="btn"
    >
      <button
        className="btnFlat btnMd btnPrimary"
        onClick={[Function]}
      >
        <img
          alt="arrow right"
          className="right"
          src="arrow.svg"
        />
      </button>
    </div>
  </div>
</div>
`;
